src/Phinx/Console/Command/AbstractCommand.php

```diff
@@ -401,4 +401,73 @@ protected function getSeedTemplateFilename()
     {
         return __DIR__ . self::DEFAULT_SEED_TEMPLATE;
     }
+
+    /**
+     * Gets environment to use for command.
+     *
+     * @return string|null name of environment, null if could not be found
+     */
+    protected function getEnvironment(InputInterface $input, OutputInterface $output): ?string
+    {
+        $environment = $input->getOption('environment');
+
+        if ($environment === null) {
+            $environment = $this->getConfig()->getDefaultEnvironment();
+            $output->writeln('<comment>warning</comment> no environment specified, defaulting to: ' . $environment);
+        } else {
+            $output->writeln('<info>using environment</info> ' . $environment);
+        }
+
+        if (!$this->getConfig()->hasEnvironment($environment)) {
+            $output->writeln(sprintf('<error>The environment "%s" does not exist</error>', $environment));
+
+            return null;
+        }
+
+        return $environment;
+    }
+
+    /**
+     * Prints out general runtime information about loaded environment.
+     *
+     * @param array|null $envOptions environment options
+     * @param \Symfony\Component\Console\Output\OutputInterface $output Output
+     * @return bool
+     */
+    protected function checkEnvironmentOptions(?array $envOptions, OutputInterface $output): bool
+    {
+        if (isset($envOptions['adapter'])) {
+            $output->writeln('<info>using adapter</info> ' . $envOptions['adapter']);
+        }
+
+        if (isset($envOptions['wrapper'])) {
+            $output->writeln('<info>using wrapper</info> ' . $envOptions['wrapper']);
+        }
+
+        if (
+            isset($envOptions['adapter'])
```

Comment:
Isn't the adapter required? Validation should early out on missing required information.

Reply:
Technically, no. It'll crash later on in this process, but right now there's no error checking when loading the config file / environment options.

Reply:
How is that not a required setting then?

Reply:
Sorry, I misspoke slightly. So you can either specify:

* `adapter` (string that's mysql, sqlite, etc.)
* `dsn` (that gets parsed into components, and will set adapter when loading config)
* `connection` (PDO object, will set adapter once Environment class is initialized which is done by the Manager when it's doing its thing after all of these prints are done).

https://github.com/cakephp/phinx/blob/a14cc9e0a9258b30f20e5c1a711a3a8c266a3c91/src/Phinx/Migration/Manager/Environment.php#L337-L344

Incidentally, the Environment class is where the lack of an adapter is finally thrown:
https://github.com/cakephp/phinx/blob/a14cc9e0a9258b30f20e5c1a711a3a8c266a3c91/src/Phinx/Migration/Manager/Environment.php#L345-L347

So, it could be that those 10 lines get moved into the Config class where the environment is loaded (and after `parseAgnosticDsn` is run) and then `adapter` would be guaranteed to be set when that block of code we're talking about here runs.

Reply:
In that case, I think we should validate them when constructing the environment. Validating and logging them in the command feels out of order.

The command can check if it has a valid environment and log then fail.

Reply:
Construction of the environment happens after all of these log statements (in the manager method that each of these methods calls).

I suppose each of these methods could call `$this->getManager()->getEnvironment()` before printing out the environment details as they do now to get that error checking. Or do some other sort of larger refactor.